pkg/mimirpb/compat.go

```diff
@@ -89,62 +89,51 @@ func (req *WriteRequest) AddHistogramSeries(lbls []labels.Labels, histograms []H
-	return *(*labels.Labels)(unsafe.Pointer(&ls))
-}
-
-// FromLabelAdaptersToLabelsWithCopy converts []LabelAdapter to labels.Labels.
-// Do NOT use unsafe to convert between data types because this function may
-// get in input labels whose data structure is reused.
-func FromLabelAdaptersToLabelsWithCopy(input []LabelAdapter) labels.Labels {
-	return CopyLabels(FromLabelAdaptersToLabels(input))
-}
-
-// CopyLabels efficiently copies labels input slice. To be used in cases where input slice
-// can be reused, but long-term copy is needed.
-func CopyLabels(input []labels.Label) labels.Labels {
-	result := make(labels.Labels, len(input))
-
-	size := 0
-	for _, l := range input {
-		size += len(l.Name)
-		size += len(l.Value)
+	builder := labels.NewScratchBuilder(len(ls))
+	for _, v := range ls {
+		builder.Add(v.Name, v.Value)
 	}
+	return builder.Labels()
+}
 
-	// Copy all strings into the buffer, and use 'yoloString' to convert buffer
-	// slices to strings.
-	buf := make([]byte, size)
-
-	for i, l := range input {
-		result[i].Name, buf = copyStringToBuffer(l.Name, buf)
-		result[i].Value, buf = copyStringToBuffer(l.Value, buf)
+// Build a labels.Labels from LabelAdaptors, with amortized zero allocations.
+func FromLabelAdaptersOverwriteLabels(builder *labels.ScratchBuilder, ls []LabelAdapter, dest *labels.Labels) {
+	builder.Reset()
+	for _, v := range ls {
+		builder.Add(v.Name, v.Value)
 	}
-	return result
+	builder.Overwrite(dest)
 }
 
-// Copies string to buffer (which must be big enough), and converts buffer slice containing
-// the string copy into new string.
-func copyStringToBuffer(in string, buf []byte) (string, []byte) {
-	l := len(in)
-	c := copy(buf, in)
-	if c != l {
-		panic("not copied full string")
+// FromLabelAdaptersToBuilder converts []LabelAdapter to labels.Builder.
+func FromLabelAdaptersToBuilder(ls []LabelAdapter, builder *labels.Builder) {
+	builder.Reset(labels.EmptyLabels())
+	for _, v := range ls {
+		builder.Set(v.Name, v.Value)
 	}
+}
 
-	return yoloString(buf[0:l]), buf[l:]
+// FromBuilderToLabelAdapters converts labels.Builder to []LabelAdapter, reusing ls.
+// Note the result may not be sorted.
+func FromBuilderToLabelAdapters(builder *labels.Builder, ls []LabelAdapter) []LabelAdapter {
+	ls = ls[:0]
+	builder.Range(func(l labels.Label) {
+		ls = append(ls, LabelAdapter{Name: l.Name, Value: l.Value})
+	})
+	return ls
 }
 
 // FromLabelsToLabelAdapters casts labels.Labels to []LabelAdapter.
-// It uses unsafe, but as LabelAdapter == labels.Label this should be safe.
-// This allows us to use labels.Labels directly in protos.
+// For now it's doing an expensive conversion: TODO figure out a faster way.
 func FromLabelsToLabelAdapters(ls labels.Labels) []LabelAdapter {
-	return *(*[]LabelAdapter)(unsafe.Pointer(&ls))
+	r := make([]LabelAdapter, 0, ls.Len())
+	ls.Range(func(l labels.Label) {
+		r = append(r, LabelAdapter{Name: l.Name, Value: l.Value})
+	})
+	return r
```

Comment:
This code seems to be used on the "hot" path. Not it seems to do an allocation per series for push requests from the ruler ([code](https://github.com/grafana/mimir/blob/bc241d994d70f78e42f8705f5ec83dda903e4026/pkg/ruler/compat.go#L87-L88)), for applying relabeling rules in the distributor ([code](https://github.com/grafana/mimir/blob/70a1cfc974b39fbfd8947d53ff9442f38eae401f/pkg/distributor/distributor.go#L786)), and once/twice for each touched timeseries by a query ([in the querier here](https://github.com/grafana/mimir/blob/1e262759d638eff986d3348a5fd4e3593f6cd39d/pkg/querier/remote_read.go#L257), [in the store-gateway here](https://github.com/grafana/mimir/blob/e16d1e397e91a43efe0c6ddb960a6ef804217961/pkg/storegateway/bucket.go#L664)). So in these cases we would hold the memory for the optimized stringlabels once and once for the unoptimized slice representation.

Have you ran any benchmarks to gauge the impact of this?

Reply:
I have run benchmarks, but the underlying code keeps changing so it is hard to keep track.
I think it's basically whack-a-mole now; we have to run it again and see if anything pops out as a hotspot.

Relabelling is already an expensive operation with regexps; the code you pointed to is [different](https://github.com/grafana/mimir/blob/labels-as-string/pkg/distributor/distributor.go#L786) in my fork.

Reply:
I'll make a start on running the Mimir benchmarks again.

Reply:
another option is to run with stringlabels and without side-by-side for some time. Some ingester zones and some store-gateway zones can run different configurations. IIRC we can also do something similar for the queriers in Grafana Labs clusters.

Is it more difficult to get two concurrent container images with and without stringlabels?

Reply:
Do I read this right that it's allocating 96% more memory?